src/protocol/modulus_conversion/mod.rs

```diff
@@ -16,17 +16,80 @@ use crate::{ff::Field, secret_sharing::replicated::semi_honest::AdditiveShare as
 ///     ...
 ///     `[ row[n].bit0, row[n].bit1, ..., row[n].bit31 ]`,
 ///  `]`
+/// `num_multi_bits`: L
 ///
 /// output:
 /// `[`
-///     `[ row[0].bit0, row[1].bit0, ..., row[n].bit0 ]`,
-///     `[ row[0].bit1, row[1].bit1, ..., row[n].bit1 ]`,
+///     `[ row[0].bit0, row[0].bit1, ..., row[0].bitL ], [ row[1].bit0, row[1].bit1, ..., row[1].bitL ], .. [ row[n].bit0, row[n].bit1, ..., row[n].bitL ]`,
+///     `[ row[0].bitL+1, ..., row[0].bit2L ], [ row[1].bitL+1, ..., row[1].bit2L ], .. [ row[n].bitL+1, ..., row[n].bit2L ], `,
 ///     ...
-///     `[ row[0].bit31, row[1].bit31, ..., row[n].bit31 ]`,
+///     `[ row[0].bitmL,  ..., row[0].bit31 ], [ row[1].bitmL, ..., row[n].bit31 ], .. [ row[n].bitmL, ..., row[n].bit31 ]`,
 /// `]`
+/// # Errors
+/// If any underlying protocol fails
+/// # Panics
+/// Panics if input doesn't have same number of bits as `num_bits`
 #[must_use]
-pub fn transpose<F: Field>(input: &[Vec<Replicated<F>>]) -> Vec<Vec<Replicated<F>>> {
-    (0..input[0].len())
-        .map(|i| input.iter().map(|b| b[i].clone()).collect::<Vec<_>>())
+pub fn split_into_multi_bit_slices<F: Field>(
+    input: &[Vec<Replicated<F>>],
+    num_bits: u32,
+    num_multi_bits: u32,
+) -> Vec<Vec<Vec<Replicated<F>>>> {
```

Comment:
same here, it feels like you should be able to borrow from `input` to get rid of the most inner vector. Is it possible to return an iterator here and cut the memory footprint to a constant value? this nested structure looks frightening when I think about 1M items input

Reply:
This can return an iterator but the consumer of output (sort) still needs a slice. In one of the places where sort is called, we need to use the input of the function in next step: https://github.com/private-attribution/ipa/blob/main/src/protocol/ipa/mod.rs#L156

Can we keep working on the iterators in the next PR?